### PR TITLE
Fixes incorrect matrix being used in ApplyGateDagger.

### PR DESCRIPTION
**Pull request by MichaelBroughton**
The controlled gate case of `ApplyGateDagger` would use the undaggered version of the gates matrix. Found to be blocking on https://github.com/tensorflow/quantum/pull/458

### CODE REVIEW

File: lib/gate_appl.h [hunk 1]
```diff
@@ -64,7 +64,7 @@ inline void ApplyGateDagger(const Simulator& simulator, const Gate& gate,
       simulator.ApplyGate(gate.qubits, matrix.data(), state);
     } else {
       simulator.ApplyControlledGate(gate.qubits, gate.controlled_by,
-                                    gate.cmask, gate.matrix.data(), state);
```
Comment (Strilanc) [inline lib/gate_appl.h:67]: Add a test.

Comment (MichaelBroughton) [reply lib/gate_appl.h:67]: Done.

File: tests/simulator_testfixture.h [hunk 2]
```diff
@@ -721,6 +721,384 @@ if __name__ == '__main__':
...
+    {-0.05464944, 0.056015424},
+    {-0.16184065, -0.009010859},
+    {0.12749553, -0.12438276},
+    {0.019615382, 0.092316},
+    {-0.04924332, 0.044155773},
+    {-0.24133444, -0.033628717},
+    {-0.18774915, 0.12311842},
+  };
+
+  for (unsigned i = 0; i < size; i++) {
```
Comment (sergeisakov) [inline tests/simulator_testfixture.h:905]: Replace `i++` with `++i` for consistency with the rest of the code?

File: tests/simulator_testfixture.h [hunk 3]
```diff
@@ -721,6 +721,384 @@ if __name__ == '__main__':
...
+    state_space.SetAmpl(state, i, final_amplitudes[i][0], final_amplitudes[i][1]);
+  }
+
+  for (int i = gates.size() - 1; i >= 0; i--) {
```
Comment (sergeisakov) [inline tests/simulator_testfixture.h:909]: Replace `i--` with `--i` for consistency with the rest of the code?

**Review by sergeisakov — APPROVED** [review]
Thank you for fixing that.

File: tests/simulator_testfixture.h (hunk 2 above)
Comment (MichaelBroughton) [reply tests/simulator_testfixture.h]: Done.

File: tests/simulator_testfixture.h (hunk 3 above)
Comment (MichaelBroughton) [reply tests/simulator_testfixture.h]: Done.

**Pull request merged**